driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -308,139 +309,66 @@ public void sendAndReceiveAsync(final CommandMessage message, final SingleResult
-                                         final ByteBufferBsonOutput bsonOutput, final String commandName, final long startTimeNanos) {
-        sendMessageAsync(bsonOutput.getByteBuffers(), message.getId(), new SingleResultCallback<Void>() {
+    private <T> void sendCommandMessageAsync(final int messageId, final Decoder<T> decoder, final SingleResultCallback<T> callback,
+                                             final ByteBufferBsonOutput bsonOutput,
+                                             final CommandEventSender commandEventSender) {
+        sendMessageAsync(bsonOutput.getByteBuffers(), messageId, new SingleResultCallback<Void>() {
             @Override
             public void onResult(final Void result, final Throwable t) {
                 bsonOutput.close();
                 if (t != null) {
-                    sendFailedEvent(startTimeNanos, message, commandName, t);
+                    commandEventSender.sendFailedEvent(t);
                     callback.onResult(null, t);
                 } else {
-                    readAsync(MessageHeader.MESSAGE_HEADER_LENGTH, new SingleResultCallback<ByteBuf>() {
+                    readAsync(MESSAGE_HEADER_LENGTH, new MessageHeaderCallback(new SingleResultCallback<ResponseBuffers>() {
                         @Override
-                        public void onResult(final ByteBuf result, final Throwable t) {
+                        public void onResult(final ResponseBuffers responseBuffers, final Throwable t) {
                             if (t != null) {
-                                sendFailedEvent(startTimeNanos, message, commandName, t);
+                                commandEventSender.sendFailedEvent(t);
+                                close();
                                 callback.onResult(null, t);
                                 return;
                             }
-                            ByteBufferBsonInput headerInputBuffer = new ByteBufferBsonInput(result);
-                            final MessageHeader messageHeader;
                             try {
-                                messageHeader = new MessageHeader(headerInputBuffer, description.getMaxMessageSize());
+                                boolean commandOk =
+                                        isCommandOk(new BsonBinaryReader(new ByteBufferBsonInput(responseBuffers.getBodyByteBuffer())));
+                                responseBuffers.reset();
+                                if (!commandOk) {
+                                    MongoException commandFailureException = getCommandFailureException(getResponseDocument(responseBuffers,
+                                            messageId, new BsonDocumentCodec()), description.getServerAddress());
+                                    commandEventSender.sendFailedEvent(commandFailureException);
+                                    throw commandFailureException;
```

Comment:
That said, we don't wrap the client's callback in error handling callback here, and never have. There is wrapping all the way up in CommandOperationHelper, which is what's protecting us now.  Maybe we should re-think at which level(s) we do that wrapping.